File: app/src/main/java/ai/elimu/appstore/room/dao/ApplicationVersionDao.java
package ai.elimu.appstore.room.dao;

import android.database.Cursor;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

import ai.elimu.appstore.room.entity.ApplicationVersion;

@Dao
public interface ApplicationVersionDao {

    @Insert
    void insert(ApplicationVersion applicationVersion);

    @Query("SELECT * FROM ApplicationVersion av WHERE av.id = :id")
    ApplicationVersion load(Long id);

    @Query("SELECT * FROM ApplicationVersion av ORDER BY av.applicationId ASC, av.versionCode DESC")
    List<ApplicationVersion> loadAll();

    @Query("SELECT * FROM ApplicationVersion av WHERE av.applicationId = :applicationId ORDER BY versionCode DESC")
    List<ApplicationVersion> loadAll(Long applicationId);

    @Update
    void update(ApplicationVersion applicationVersion);

    @Query("DELETE FROM ApplicationVersion WHERE applicationId = :applicationId")
    void delete(Long applicationId);
}

File: app/src/main/java/ai/elimu/appstore/room/dao/ApplicationVersionDao.kt
package ai.elimu.appstore.room.dao

import ai.elimu.appstore.room.entity.ApplicationVersion
import androidx.room.Dao
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Update

@Dao
interface ApplicationVersionDao {
    @Insert
    fun insert(applicationVersion: ApplicationVersion)

    @Query("SELECT * FROM ApplicationVersion av WHERE av.id = :id")
    fun load(id: Long?): ApplicationVersion?

    @Query("SELECT * FROM ApplicationVersion av ORDER BY av.applicationId ASC, av.versionCode DESC")
    fun loadAll(): List<ApplicationVersion>

    @Query("SELECT * FROM ApplicationVersion av WHERE av.applicationId = :applicationId ORDER BY versionCode DESC")
    fun loadAll(applicationId: Long?): List<ApplicationVersion>

    @Update
    fun update(applicationVersion: ApplicationVersion)

    @Query("DELETE FROM ApplicationVersion WHERE applicationId = :applicationId")
    fun delete(applicationId: Long)
}
